cql3/cql_value.hh

```diff
@@ -0,0 +1,207 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Copyright (C) 2021-present ScyllaDB
+ *
+ * Modified by ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "cql3/values.hh"
+
+namespace cql3 {
+    // A value represented by empty bytes()
+    // Created for example by using blobasint(0x) in cqlsh
+    struct empty_value {
+    };
+
+    struct bool_value {
+        bool value;
+    };
+
```

Comment:
Why not use `data_value`? Or alternatively `managed_bytes`+`data_type`?

`data_value` is a {type, unserialized value} pair, and `managed_bytes`+`data_type` is the same but serialized. Probably the latter will fit better with the rest of the system.

Reply:
I was following the representation from `terminal`.
In uses of `terminal` we need to do four things:
* Convert value to serialized bytes (old `get`)
* Convert serialized bytes to value (old `value::from_serialized`)
* Sometimes access the deserialized value of a simple value (limit, timeout, ...)
* Access elements of a collection, but only one level deep (like in [`lists::do_append`](https://github.com/scylladb/scylla/blob/3633c077bed2662a2e66e6e2034856cc6df74923/cql3/lists.cc#L369))

> Why not `data_value`

In case of a `list<frozen<list<int>>` a `data_value` would keep `vector<vector<int>>` + many smart pointers, which is not needed, we just need `vector<managed_bytes>`.
`data_value` would work, but I'm a bit worried about performance - a list of 100 ints would require 100 `unique_ptr` allocations and 100 `shared_ptr` copies :/

> Why not `managed_bytes`+`data_type`?

With collections, going from `managed_bytes` to `vector<managed_bytes>` [is not trivial](https://github.com/scylladb/scylla/blob/3633c077bed2662a2e66e6e2034856cc6df74923/cql3/lists.cc#L127). We would need some conversion functions to convert from and to the representation that we need. The representation we need is more or less `cql_value` and there are conversion functions: `from_serialized`, `to_raw_value`.

Reply:
> I was following the representation from `terminal`.

Incremental changes are best, so we can do that, to avoid making the change too complicated to ever land. Still, we should discuss future changes too.

> In uses of `terminal` we need to do four things:
> 
> * Convert value to serialized bytes (old `get`)

That's the main use-case, so we should optimize for that.

> * Convert serialized bytes to value (old `value::from_serialized`)

If we store as serialized (bytes_opt or managed_bytes_opt), then this becomes

if (using same cql serialization format) {
    copy
} else {
    deserialize and reserialize
}

which is efficient (since no one uses the old serialization format, I'll start the process to phase it out)

> * Sometimes access the deserialized value of a simple value (limit, timeout, ...)

Simple values are cheap to deserialize, so I think this is fine.

> * Access elements of a collection, but only one level deep (like in [`lists::do_append`](https://github.com/scylladb/scylla/blob/3633c077bed2662a2e66e6e2034856cc6df74923/cql3/lists.cc#L369))


I think this is addressed by the partially_deserialize_*() function family. At least do_append() doesn't need random access.

> 
> > Why not `data_value`
> 
> In case of a `list<frozen<list<int>>` a `data_value` would keep `vector<vector<int>>` + many smart pointers, which is not needed, we just need `vector<managed_bytes>`.
> `data_value` would work, but I'm a bit worried about performance - a list of 100 ints would require 100 `unique_ptr` allocations and 100 `shared_ptr` copies :/

I agree it's a bad idea.

> 
> > Why not `managed_bytes`+`data_type`?
> 
> With collections, going from `managed_bytes` to `vector<managed_bytes>` [is not trivial](https://github.com/scylladb/scylla/blob/3633c077bed2662a2e66e6e2034856cc6df74923/cql3/lists.cc#L127). We would need some conversion functions to convert from and to the representation that we need. The representation we need is more or less `cql_value` and there are conversion functions: `from_serialized`, `to_raw_value`.

I think partially_deserialize() works for that. If not, we can consider different expression subtypes with different representations, depending on what the caller prefer.